### PR TITLE
[FLINK-17126] [table-planner] Correct the execution behavior of BatchTableEnvironment

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*In previous versions, BatchTableEnvironment.execute() can both trigger table and DataSet programs. For example:
ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
BatchTableEnvironment tableEnv = BatchTableEnvironment.create(env, new TableConfig());
tEnv.sqlUpdate("INSERT INTO sink1 SELECT a, b FROM MyTable1");
Table table = tEnv.sqlQuery("SELECT c, d from MyTable2");
DataSet<Row> dataSet = tEnv.toDataSet(table, Row.class);
dataSet...
env.execute("job name");
// or tEnv.execute("job name");
// both env.execute and tEnv.execute can trigger the execution

To correct current messy trigger point, we propose that: table programs can only be triggered by BatchTableEnvironment.execute(). Once table program is converted into DataSet program (through toDataSet() method), it can only be triggered by ExecutionEnvironment.execute()
The example with new behavior:
ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
BatchTableEnvironment tableEnv = BatchTableEnvironment.create(env, new TableConfig());
tEnv.sqlUpdate("INSERT INTO sink1 SELECT a, b FROM MyTable1");
Table table = tEnv.sqlQuery("SELECT c, d from MyTable2");
DataSet<Row> dataSet = tEnv.toDataSet(table, Row.class);
dataSet...
env.execute("job name") ; // only trigger the DataSet program
tEnv.execute("job name"); // only trigger the sql update query program

This issue is similar to FLINK-16363.*

## Brief change log

  - *Introduce CollectionPipelineExecutor for testing purpose through Executor*
  - *set `execution.target` and `execution.attached` in TestEnvironment for testing purpose through Executor*
  - *Correct the execution behavior of BatchTableEnvironment*

## Verifying this change

This change added tests and can be verified as follows:
- *Added CollectionExecutorTest to validate CollectionPipelineExecutor*
- *Extended TableEnvironmentITCase to verify new execution behavior*
  

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/util/DummyNoOpOperator.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.io.IOException;
+
+/**
+ * This is dummy {@link NoOpOperator}, which context is {@link DummyExecutionEnvironment}.
+ */
+public class DummyNoOpOperator<IN> extends NoOpOperator<IN> {
+
+	public DummyNoOpOperator(
+		ExecutionEnvironment dummyExecEnv,
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/util/DummyNoOpOperator.java:35]: Indent.

**Comment by wuchong** [comment]
Could you also add a release note in the JIRA issue?

**Comment by godfreyhe** [comment]
> Could you also add a release note in the JIRA issue?

thanks for the suggestion @wuchong , I updated the PR and the release note.

**Comment by kl0u** [comment]
Hi @wuchong and @godfreyhe, could you elaborate on the added functionality of this PR? 

The `CollectionEnvironment` is only intended for testing and only for batch. This is the reason why we did not add an `executor` when executors were originally introduced. Do you have cases where this is actually needed by users?

**Comment by godfreyhe** [comment]
hi @kl0u `CollectionEnvironment` is used in flink-table-planner for testing, e.g. `org.apache.flink.table.runtime.batch.sql.TableEnvironmentITCase#testInsertIntoMemoryTable`
and `org.apache.flink.table.runtime.batch.table.TableSinkITCase#testOutputFormatTableSink`. 
User could also use a collection as source for demo.
After we use executor to submit a job, an executor for `CollectionEnvironment` is also needed.

**Comment by wuchong** [comment]
Hi @kl0u , we refactored `BatchTableEnviroment` to use `executor` to submit the Pipline. That's why we need an executor for `CollectionEnvironment` because tests are using `CollectionEnvironment`.

Sorry for didn't ping you to review this part.

**Comment by kl0u** [comment]
Hi @godfreyhe and @wuchong , thanks for letting me know. I think that using the `CollectionEnvironment` is not a good practice, as it is not going through all the steps that a regular user program would go. How about changing the tests to use `LocalEnvironment`, or a more realistic environment in general. 

Another reason I am asking is because there were informal discussions about getting rid of the `CollectionEnvironment` altogether at some point.

What is your opinion @twalthr on this?

**Comment by twalthr** [comment]
The collection mode is very old code. We used it in table tests for speeding up ITCases back in the days when build times were limited by Travis. I think nobody is really using the collection mode. How about we just update the test bases and use cluster mode everywhere instead of adding more legacy?

**Comment by godfreyhe** [comment]
@kl0u @twalthr Thanks for the feedback, I think it's time to remove `CollectionEnvironment` usage from flink-table tests, I will create a JIRA to do that, and I will also revert the commit "[FLINK-17126][java] Introduce CollectionPipelineExecutor for CollectionEnvironment (c983ac9c)" to avoid someone using `CollectionEnvironment` for testing or demo. (if someone use `CollectionEnvironment`, he will get `java.lang.NullPointerException: No execution.target specified in your configuration file.`)  WDYT ?

**Comment by kl0u** [comment]
@godfreyhe I think that this is a valid tradeoff for now. In the future we may have an exception that says that `CollectionEnviroment not allowed to be used with TableEnvironment. Please use XXXX instead`. For `DataSet` we do not have this problem because the execution logic is hardcoded in the environment.

**Comment by godfreyhe** [comment]
Thanks @kl0u , I created a JIRA to track this. https://issues.apache.org/jira/browse/FLINK-17383

**Comment by kl0u** [comment]
Hi @godfreyhe , from the discussion here I got that the `CollectionPipelineExecutor` would be removed but this is still in the codebase and it is also part of the help message when using `-h`.

Is this intentional? And, if yes, what are the plans for removing it?

**Pull request closed without merging**